src/image/detail/MaskDict.cc

```diff
@@ -31,18 +32,48 @@ class MaskDict::GlobalState {
 
     using HandleList = std::set<std::weak_ptr<MaskDict>, std::owner_less<std::weak_ptr<MaskDict>>>;
 
-    static GlobalState & get() {
-        static GlobalState instance;
-        return instance;
-    }
+    /*
+     * RAII guard for GlobalState.
+     *
+     * Default-constructing a Lock instance is the only way to access
+     * GlobalState - this acquires GlobalState's mutex, ensuring only one
+     * thread has access to its attributes at one time.  The mutex is released
+     * when the Lock instance goes out of scope.
+     */
+    class Lock {
+    public:
+
+        Lock() :
+            allMaskDicts(_get()._allMaskDicts),
+            defaultMaskDict(_get()._defaultMaskDict),
+            _guard(_get()._mutex)
```

Comment:
That could work (though it would make `GlobalState's` singleton status confusing in much the same way as my earlier remark that the class is not itself thread-safe). I was thinking of a design where `Lock` contains only the `lock_guard`, and the instance `get` method calls the existing `GlobalState::_get()`.